storage/tests/system.py

```diff
@@ -1470,3 +1470,150 @@ def test_bucket_lock_retention_policy(self):
+        blob_acl.all().grant_read()
+        with self.assertRaises(exceptions.BadRequest):
+            blob_acl.save()
+
+    @unittest.expectedFailure  # Until rollout completes, ETA 2019-01-10
+    def test_existing_bucket_set_bpo(self):
+        import requests
+
+        new_bucket_name = "set-bpo" + unique_resource_id("-")
+        self.assertRaises(
+            exceptions.NotFound, Config.CLIENT.get_bucket, new_bucket_name
+        )
+        bucket = retry_429(Config.CLIENT.create_bucket)(new_bucket_name)
+        self.case_buckets_to_delete.append(new_bucket_name)
+
+        blob_name = "my-blob.txt"
+        blob = bucket.blob(blob_name)
+        payload = b"DEADBEEF"
+        blob.upload_from_string(payload)
+
+        blob.make_public()
+
+        ok_response = requests.get(blob.public_url)
+        self.assertEqual(ok_response.content, payload)
+
+        # XXX: Work around 400 error: "Cannot disable object policies because
+        # default object acls are not empty."
+        response = Config.CLIENT._connection.api_request(
+            method="PATCH",
+            path=bucket.path,
+            data={},
+            query_params={'predefinedDefaultObjectAcl': 'private'},
+            _target_object=bucket,
+        )
+
+        bucket.iam_configuration.bucket_policy_only = True
+        bucket.patch()
+
+        failed_response = requests.get(blob.public_url)
+        #XXX: the request for downloading the object via its public URL does
+        #     *not* fail as of 2019-01-08T18:43:00Z
+        self.assertEqual(failed_response.status_code, 403)
+
+    def test_bpo_set_unset_preserves_acls(self):
+        new_bucket_name = "bpo-acls" + unique_resource_id("-")
+        self.assertRaises(
+            exceptions.NotFound, Config.CLIENT.get_bucket, new_bucket_name
+        )
+        bucket = retry_429(Config.CLIENT.create_bucket)(new_bucket_name)
+        self.case_buckets_to_delete.append(new_bucket_name)
+
+        blob_name = "my-blob.txt"
+        blob = bucket.blob(blob_name)
+        payload = b"DEADBEEF"
+        blob.upload_from_string(payload)
+        blob_policy = blob.get_iam_policy()
+
+        # Preserve ACLs before setting BPO
+        bucket_acl_before = list(bucket.acl)
+        blob_acl_before = list(bucket.acl)
+
+        # XXX: Work around 400 error: "Cannot disable object policies because
+        # default object acls are not empty."
+        response = Config.CLIENT._connection.api_request(
+            method="PATCH",
+            path=bucket.path,
+            data={},
+            query_params={'predefinedDefaultObjectAcl': 'private'},
+            _target_object=bucket,
+        )
+
+        # Set BPO
+        bucket.iam_configuration.bucket_policy_only = True
+        bucket.patch()
+
+        # While BPO is set, cannot get / set ACLs
+        with self.assertRaises(exceptions.BadRequest):
+            bucket.acl.reload()
+
+        # XXX This should raise, but doesn't as of 2018-01-08T19:00:00Z
```

Comment:
What is the difference between the http requests being sent as a result of bucket.acl.reload() vs bucket.acl.clear()? What endpoints are being used and what's in the request body?

Reply:
@chestercun `bucket.acl.clear()` is a POST request with an empty ACL list.  `bucket.acl.get()` is a GET request.

Reply:
The only bucket ACL endpoint that uses POST is https://www.googleapis.com/storage/v1/b/[BUCKET_NAME]/acl 
and that API is only for insert a new ACL entry onto an existing ACL. I don't still understand how bucket.acl.clear() is supposed to work.

 https://cloud.google.com/storage/docs/json_api/v1/bucketAccessControls/insert

Reply:
@chestercun Sorry, my finger-memory supplied POST rather than PATCH:  it uses the [`bucketAccessControls.patch`](https://cloud.google.com/storage/docs/json_api/v1/bucketAccessControls/patch) API.

Reply:
Does this API actually delete an ACL entry on a non-BPO bucket? I'm not sure why DELETE wouldn't be used to remove an ACL entry instead.

Reply:
We don't expose API operations on individual ACL entries:  the PATCH request updates the entire ACL: for `bucket.acl.clear()` it removes all entries (that the server allows).  `bucket.acl.save()` replaces the bucket's ACL entries with those on the local object (also using PATCH).